### PR TITLE
Disable part files for OC ext storage backend + s2s backend

### LINKED ISSUE #13157: Server two server sharing + sync client + encryption on both sides fails
### Steps to reproduce
1. Setup ownCloud A with encryption
2. Setup ownCloud B with encryption
3. On B, create a folder "test"
4. Share "test" with link and allow upload
5. Mount the link share in ownCloud A
6. Setup the sync client for A
7. Upload a file into "test" using the sync client

### Expected result
Successful upload

### Actual result
"expected filesize X got Y".
B returns the encrypted file size for the part file when asked.

See https://github.com/owncloud/core/pull/13156#issuecomment-69052374 for a similar case.

This ticket is to document the case and evaluate severity.

CC @schiesbn @DeepDiver1975 @karlitschek 

The fix proposed in https://github.com/owncloud/core/pull/13156#issuecomment-69057985 should work, which uploads the part file directly as the final name instead of doing "upload part file + rename", since the backend OC already has the part file logic.

I'll submit a PR for it that will cover both server to server sharing and mounting OC as backend.

### PR DESCRIPTION
**Pull request by PVince81**
When uploading files to an OC ext storage backend or when using server
to server sharing storage, part files aren't needed because the backend
already has its own part files and takes care of the final atomic rename
operation.

This also fixes issues when using two encrypted ownCloud instances where
one mounts the other either as external storage (ownCloud backend) or
through server to server sharing.

Note: this is a quick fix for now. In the future we'll look into providing a method on the Storage class and/or having storage specific storage attributes/chunk handlers that would let the storage class decide whether or not part files/chunks/whatever is required or not.

Please review @schiesbn @icewind1991 @DeepDiver1975 

Fixes https://github.com/owncloud/core/issues/13157 and https://github.com/owncloud/core/issues/5952
To test: setup two ownCloud instances with BOTH having encryption enabled.
#### Server to server sharing

Mount OC B in OC A.
Setup sync client for OC A.
- [x] upload small file (no chunking)
- [x] upload big file (chunking)
#### External storage backend "ownCloud"

Mount OC B in OC A.
Setup sync client for OC A.
- [x] upload small file (no chunking)
- [x] upload big file (chunking)
#### Regression testing

Uploading to regular storage or another one (ex: SFTP):
- [x] upload small file (no chunking)
- [x] upload big file (chunking)

I tested this and it worked.

### CODE REVIEW

File: lib/private/connector/sabre/file.php [hunk 1]
```diff
@@ -74,8 +74,16 @@ public function put($data) {
 			return $this->createFileChunked($data);
 		}
 
-		// mark file as partial while uploading (ignored by the scanner)
-		$partFilePath = $this->path . '.ocTransferId' . rand() . '.part';
+		list($storage, $internalPath) = $this->fileView->resolvePath($this->path);
```
Comment (LukasReschke) [inline lib/private/connector/sabre/file.php]: Is `$internalPath` actually needed here for anything except eating memory? - I doubt it, remove it?

File: lib/private/connector/sabre/file.php [hunk 2]
```diff
@@ -267,23 +277,31 @@ private function createFileChunked($data)
 		}
 
 		if ($chunk_handler->isComplete()) {
+			list($storage, $internalPath) = $this->fileView->resolvePath($path);
```
Comment (LukasReschke) [inline lib/private/connector/sabre/file.php]: Is `$internalPath` actually needed here for anything except eating memory? - I doubt it, remove it?

File: lib/private/connector/sabre/file.php [hunk 3]
```diff
@@ -267,23 +277,31 @@ private function createFileChunked($data)
...
-					// only delete if an error occurred and the target file was already created
-					if ($fileExists) {
-						$this->fileView->unlink($targetPath);
+				if ($needsPartFile) {
+					// we first assembly the target file as a part file
+					$partFile = $path . '/' . $info['name'] . '.ocTransferId' . $info['transferid'] . '.part';
+					$chunk_handler->file_assemble($partFile);
+
+					// here is the final atomic rename
+					$targetPath = $path . '/' . $info['name'];
```
Comment (LukasReschke) [inline lib/private/connector/sabre/file.php]: `$targetPath` is also accessed in L310 and might be undefined in the execution. Probably was there already before though too.

Comment (DeepDiver1975) [inline lib/private/connector/sabre/file.php]: > Probably was there already before though too.

No - introduced with this patch

File: lib/private/connector/sabre/file.php [hunk 4]
```diff
@@ -303,4 +320,19 @@ private function createFileChunked($data)
...
+
+	/**
+	 * Returns whether a part file is needed for the given storage
+	 * or whether the file can be assembled/uploaded directly on the
+	 * target storage.
+	 *
+	 * @param \OCP\Files\Storage $storage storage to check
+	 * @param bool true if the storage needs part file handling
+	 */
+	private function needsPartFile($storage) {
```
Comment (LukasReschke) [inline lib/private/connector/sabre/file.php]: Enforce type? 
```php
private function needsPartFile(\OCP\Files\Storage $storage) {
```

**Comment by PVince81** [comment]
Anything else to change ?

**Comment by DeepDiver1975** [comment]
looks good from my pov :+1:

**Comment by DeepDiver1975** [comment]
> Fixes #13157 and #5952

https://github.com/owncloud/core/issues/5952 is scheduled for 8.1 - we shall move it to 8.0 I assume?

**Comment by PVince81** [comment]
Yes. We set it to 8.1 back then because the scope of the fix wasn't clear.

File: lib/private/connector/sabre/file.php (hunk 4 above)
Comment (LukasReschke) [inline lib/private/connector/sabre/file.php]: Mhm – generally speaking a type annotation in the PHPDoc is worth nothing when not enforced as well :-)

But go whatever you feel safer here :-)

**Comment by PVince81** [comment]
@MorrisJobke @LukasReschke @icewind1991 can we merge this ?

**Comment by MorrisJobke** [comment]
I will review this one.

**Comment by MorrisJobke** [comment]
Tested with small (text file) and big 700 MB video. Both work after the fix (via web UI and sync client)

**Comment by MorrisJobke** [comment]
:+1:

**Pull request merged**